packages/opentelemetry-plugin-fetch/src/fetch.ts

```diff
@@ -0,0 +1,350 @@
+
+// how long to wait for observer to collect information about resources
+// this is needed as event "load" is called before observer
+// hard to say how long it should really wait, seems like 300ms is
+// safe enough
+const OBSERVER_WAIT_TIME_MS = 300;
+
+/**
+ * FetchPlugin Config
+ */
+export interface FetchPluginConfig extends api.PluginConfig {
+  // the number of timing resources is limited, after the limit
+  // (chrome 250, safari 150) the information is not collected anymore
+  // the only way to prevent that is to regularly clean the resources
+  // whenever it is possible, this is needed only when PerformanceObserver
+  // is not available
+  clearTimingResources?: boolean;
+  // urls which should include trace headers when origin doesn't match
+  propagateTraceHeaderCorsUrls?: web.PropagateTraceHeaderCorsUrls;
+}
+
+/**
+ * This class represents a fetch plugin for auto instrumentation
+ */
+export class FetchPlugin extends core.BasePlugin<Promise<Response>> {
+  moduleName = 'fetch';
+  private _usedResources = new WeakSet<PerformanceResourceTiming>();
+  private _tasksCount = 0;
+
+  constructor(protected _config: FetchPluginConfig = {}) {
+    super('@opentelemetry/plugin-fetch', VERSION);
+  }
+
+  /**
+   * Add cors pre flight child span
+   * @param span
+   * @param corsPreFlightRequest
+   */
+  private _addChildSpan(
+    span: api.Span,
+    corsPreFlightRequest: PerformanceResourceTiming
+  ): void {
+    this._tracer.withSpan(span, () => {
+      const childSpan = this._tracer.startSpan('CORS Preflight', {
+        startTime: corsPreFlightRequest[web.PerformanceTimingNames.FETCH_START],
+      });
+      web.addSpanNetworkEvents(childSpan, corsPreFlightRequest);
+      childSpan.end(
+        corsPreFlightRequest[web.PerformanceTimingNames.RESPONSE_END]
+      );
+    });
+  }
+
+  /**
+   * Adds more attributes to span just before ending it
+   * @param span
+   * @param response
+   */
+  private _addFinalSpanAttributes(
+    span: api.Span,
+    response: FetchResponse
+  ): void {
+    const parsedUrl = web.parseUrl(response.url);
+    span.setAttribute(AttributeNames.HTTP_STATUS_CODE, response.status);
+    span.setAttribute(AttributeNames.HTTP_STATUS_TEXT, response.statusText);
+    span.setAttribute(AttributeNames.HTTP_HOST, parsedUrl.host);
+    span.setAttribute(
+      AttributeNames.HTTP_SCHEME,
+      parsedUrl.protocol.replace(':', '')
+    );
+    span.setAttribute(AttributeNames.HTTP_USER_AGENT, navigator.userAgent);
+  }
+
+  /**
+   * Add headers
+   * @param options
+   * @param spanUrl
+   */
+  private _addHeaders(options: RequestInit, spanUrl: string): void {
+    if (
+      !web.shouldPropagateTraceHeaders(
+        spanUrl,
+        this._config.propagateTraceHeaderCorsUrls
+      )
+    ) {
+      return;
+    }
+    const headers: { [key: string]: unknown } = {};
+    api.propagation.inject(headers);
+    options.headers = Object.assign({}, headers, options.headers || {});
```

Comment:
Because doing like this you might override the headers that user directly added. Secondly if options.headers is null, then you will inject new headers into new object ({}) not into options.headers

Reply:
I'm sorry it would have been better written like this

```suggestion
    options.headers = Object.assign({}, options.headers);
    api.propagation.inject(options.headers);
```

Reply:
but you will still override the user headers if for any reason someone defines the header with the same name.
for example 
```
options.headers = {x-b3-traceid: 'some trace id defined by user'}
```
and after this
options.headers = {x-b3-traceid: 'trace id injected from propagation'}

Reply:
I see. It's fine like this